### PR TITLE
[QoS] Support generating QoS test parameters according to configuration on Mellanox platform

### PR DESCRIPTION
**Pull request by stephenxs**
### Description of PR

Summary:
1. Add a plugin that generates QoS test parameters according to configuration for the Mellanox platform.
   The plugin will be load fixture dutQosConfig
2. Sometimes a deviation can be introduced by calculating test parameters automatically.
   Support passing margin from the qos_params for the purpose of covering the deviation.
3. Support passing packet size from the qos_params for lossy queue and queue shared watermark test.

Signed-off-by: Stephen Sun <stephens@mellanox.com>

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### CODE REVIEW

File: tests/qos/files/mellanox/qos_param_generator.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
+import math
+
+class QosParamMellanox(object):
+    def __init__(self, qos_params, asictype, speed_cablelen, ingressLosslessProfile, ingressLossyProfile, egressLosslessProfile, egressLossyProfile):
```
Comment (keboliu) [inline tests/qos/files/mellanox/qos_param_generator.py:4]: speed_cablelen --> speed_cable_len

File: tests/qos/files/mellanox/qos_param_generator.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        asic_param_dic = {
+            'spc1': {
+                'cellsize': 96,
```
Comment (keboliu) [inline tests/qos/files/mellanox/qos_param_generator.py:7]: cellsize --> cell_size

File: tests/qos/qos_sai_base.py [hunk 3]
```diff
@@ -529,10 +529,27 @@ def dutQosConfig(self, duthost, ingressLosslessProfile):
             if vendorAsic in hostvars.keys() and mgFacts["minigraph_hwsku"] in hostvars[vendorAsic]:
                 dutAsic = asic
                 break
+
         assert dutAsic, "Cannot identify DUT ASIC type"
 
+        if isMellanoxDevice(duthost):
```
Comment (keboliu) [inline tests/qos/qos_sai_base.py:535]: isMellanoxPlatform?  And, shall we decide the platform type by get platform name from config_db METADATA?

File: tests/qos/files/qos.yml [hunk 4]
```diff
@@ -17,140 +17,130 @@
 #   xoff_1 for 50G
 #   xoff_2 for 100G
 qos_params:
-    spc1:
-        40000_5m:
+    mellanox:
+        profile:
             pkts_num_leak_out: 0
```
Comment (keboliu) [inline tests/qos/files/qos.yml:22]: why needed here?

File: tests/qos/files/qos.yml [hunk 5]
```diff
@@ -17,140 +17,130 @@
...
             dscp: 8
             ecn: 1
             pg: 0
-            pkts_num_trig_egr_drp: 67965
+            pkts_num_leak_out: 0
+            packet_size: 300
+            pkts_num_margin: 4
+        wrr:
+            ecn: 1
+            q0_num_of_pkts: 140
```
Comment (keboliu) [inline tests/qos/files/qos.yml:86]: why different with the original values?

File: tests/saitests/sai_qos_tests.py [hunk 6]
```diff
@@ -1611,12 +1639,17 @@ def runTest(self):
         pkts_num_fill_min = int(self.test_params['pkts_num_fill_min'])
         pkts_num_trig_drp = int(self.test_params['pkts_num_trig_drp'])
         cell_size = int(self.test_params['cell_size'])
+        if 'packet_size' in self.test_params.keys():
+            default_packet_length = int(self.test_params['packet_size'])
```
Comment (keboliu) [inline tests/saitests/sai_qos_tests.py:1643]: looks weird to me, I think need to add a "packet_length", it either get from the test_param or uses "default_packet_length" 

pkt = simple_tcp_packet(pktlen=default_packet_length...)   ---> pkt = simple_tcp_packet(pktlen=packet_length...)

File: tests/qos/files/mellanox/qos_param_generator.py (hunk 1 above)
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py]: Fixed.

File: tests/qos/files/mellanox/qos_param_generator.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/files/mellanox/qos_param_generator.py]: Fixed.

File: tests/qos/files/qos.yml (hunk 4 above)
Comment (stephenxs) [reply tests/qos/files/qos.yml:22]: This is the frame work for all the generic mellanox parameters. The plugin will generate the parameters which depends on the qos configuration.

File: tests/qos/files/qos.yml (hunk 5 above)
Comment (stephenxs) [reply tests/qos/files/qos.yml:86]: This is the value originally our internal testcase used.

File: tests/saitests/sai_qos_tests.py (hunk 6 above)
Comment (stephenxs) [reply tests/saitests/sai_qos_tests.py:1643]: There is only one packet size in this function. 
So I think we don't need to use the "default_" prefix. 
Neither do we need to introduce two variables (default_packet_length, packet_length).
Just name it as packet_length is ok.

Comment (stephenxs) [reply tests/saitests/sai_qos_tests.py]: For all functions where the packet size is fetched from test parameters, change `default_packet_length` to `packet_length`

**Pull request closed without merging**